core/src/main/scala/kafka/log/Log.scala

```diff
@@ -470,22 +470,44 @@ class Log(val dir: File,
     trace("Reading %d bytes from offset %d in log %s of length %d bytes".format(maxLength, startOffset, name, size))
 
-    // check if the offset is valid and in range
-    val next = nextOffsetMetadata.messageOffset
-    if(startOffset == next)
-      return FetchDataInfo(nextOffsetMetadata, MessageSet.Empty)
-    
+    // We first create a few local variables to avoid race conditions with updates to the log. The order matters here.
+    // We can allow currentLastEntry to have a newer value than currentNextOffsetMetadata but not vice versa.
+    val currentNextOffsetMetadata = nextOffsetMetadata
+    val currentNext = currentNextOffsetMetadata.messageOffset
+    val currentLastEntry = segments.lastEntry
+
+    if(startOffset == currentNext)
+      return FetchDataInfo(currentNextOffsetMetadata, MessageSet.Empty)
+
     var entry = segments.floorEntry(startOffset)
-      
+
     // attempt to read beyond the log end offset is an error
-    if(startOffset > next || entry == null)
-      throw new OffsetOutOfRangeException("Request for offset %d but we only have log segments in the range %d to %d.".format(startOffset, segments.firstKey, next))
+    if(startOffset > currentNext || entry == null)
+      throw new OffsetOutOfRangeException(("Request for offset %d but we only have log segments in the range %d " +
+        "to %d.").format(startOffset, segments.firstKey, currentNext))
     
-    // do the read on the segment with a base offset less than the target offset
+    // Do the read on the segment with a base offset less than the target offset
     // but if that segment doesn't contain any messages with an offset greater than that
     // continue to read from successive segments until we get some messages or we reach the end of the log
+    // Note that at this point startOffset < currentNext. There are three cases:
+    // 1. No new log segment is created after we created local variable currentNextOffsetMetadata and currentLastEntry
+    // 2. A new log segment is created after we created local variable currentNextOffsetMetadata but before we created
+    //    currentLastEntry. i.e currentLastEntry is actually newer than currentNextOffsetMetadata
+    // 3. A new log segment is created after we created currentNextOffsetMetadata and currentLastEntry.
+    // In all three cases, reading will start from currentLastEntry or older segments, but not the newly rolled segment.
+    // In case (2) and (3), if the reading starts from the currentLastEntry, it might be reading a little conservatively
+    // because we capped the reading up to currentNextOffsetMetadata.relativePositionInSegment even though the log
+    // segment being read is no longer the active log segment. But it is fine because later reading will see updated
+    // messages.
     while(entry != null) {
-      val fetchInfo = entry.getValue.read(startOffset, maxOffset, maxLength)
+      val maxPosition = {
+        if (entry == currentLastEntry) {
```

Comment:
@junrao In the case you mentioned, the last message in L1 should not be compacted because it will be the key of the largest offset. If so, as long as the startOffset used to fall into L1, it will still be in L1. Are you saying if that last message is a tumbstone then we will have this issue? But the tombstone itself must also be deleted, so LogCleanerDeleteRetentionMs needs to be very small. It can happen but seem with extremely small chance given we only use tombstone for offset topic and default LogCleanerDeleteRetentionMs is high.

I can change the code back to use the two phase checks. That is probably safer though less clean.